src/picongpu/include/plugins/particleMerging/ParticleMerger.kernel

```diff
@@ -0,0 +1,422 @@
+/* Copyright 2017 Heiko Burau
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "plugins/particleMerging/VoronoiCell.hpp"
+#include "particles/access/Cell2Particle.hpp"
+#include "memory/Array.hpp"
+#include "memory/IndexPool.hpp"
+
+namespace picongpu
+{
+namespace plugins
+{
+namespace particleMerging
+{
+
+using namespace PMacc;
+
+
+/** Implements a particle merging algorithm based on
+ *
+ * Luu, P. T., Tückmantel, T., & Pukhov, A. (2016).
+ * Voronoi particle merging algorithm for PIC codes.
+ * Computer Physics Communications, 202, 165-174.
+ *
+ * \tparam T_ParticlesBox container of the particle species
+ */
+template< class T_ParticlesBox >
+struct ParticleMergerKernel
+{
+    using ParticlesBox = T_ParticlesBox;
+    using FramePtr = typename ParticlesBox::FramePtr;
+    using FrameType = typename ParticlesBox::FrameType;
+    using ArrayVoronoiCells = memory::Array<
+        VoronoiCell,
+        MAX_VORONOI_CELLS
+    >;
+    using VoronoiIndexPool = memory::IndexPool<
+        int32_t,
+        MAX_VORONOI_CELLS
+    >;
+
+
+    ParticlesBox particlesBox;
+    uint32_t minParticlesToMerge;
+    float_X posSpreadThreshold2;
+    float_X absMomSpreadThreshold;
+    float_X relMomSpreadThreshold;
+    float_X minMeanEnergy;
+
+    ParticleMergerKernel(
+        ParticlesBox particlesBox,
+        uint32_t minParticlesToMerge,
+        float_X posSpreadThreshold,
+        float_X absMomSpreadThreshold,
+        float_X relMomSpreadThreshold,
+        float_X minMeanEnergy
+    ) :
+        particlesBox( particlesBox ),
```

Comment:
Note: It would be more save to to use `m_VARNAME` for the local member to avoid that variables are shadowed (can be checked with the compiler flag `-Wshadow`). This was a source of many bugs in the past.